### PR TITLE
feat: Calculate all system tables "on demand"

### PR DESCRIPTION
**Pull request by alamb**
Fixes https://github.com/influxdata/influxdb_iox/issues/1436

# Rationale
See explanation on https://github.com/influxdata/influxdb_iox/pull/1446. TLDR is that the contents of the system table are computed many many times when they are not actually being read.

# Changes
Port `system.chunk_columns`, `system.columns` and `system.operations` so it computes data only when actually needed and only `Arc::clone`s stuff on common operations

# PR sequence
- [x] system.chunks (https://github.com/influxdata/influxdb_iox/pull/1446)
- [x] system.chunk_columns (this PR)
- [x] system.columns (this PR)
- [x] system.operations (this PR)

### CODE REVIEW

File: server/src/db/system_tables.rs [hunk 1]
```diff
@@ -80,32 +97,58 @@ impl SchemaProvider for SystemSchemaProvider {
...
+            COLUMNS => Some(Arc::clone(&self.columns)),
+            CHUNK_COLUMNS => Some(Arc::clone(&self.chunk_columns)),
+            OPERATIONS => Some(Arc::clone(&self.operations)),
+            _ => None,
         }
+    }
+}
+
+/// The minimal thing that a system table needs to implement
+trait IoxSystemTable: Send + Sync {
```
Comment (alamb) [inline server/src/db/system_tables.rs:111]: The basica pattern is to make a struct that implements this trait for each of the system tables

File: server/src/db/system_tables.rs [hunk 2]
```diff
@@ -358,24 +463,69 @@ fn assemble_chunk_columns(
...
+        schema,
+        vec![
             Arc::new(partition_key.finish()) as ArrayRef,
-            false,
-        ),
-        ("chunk_id", Arc::new(chunk_id.finish()), false),
-        ("table_name", Arc::new(table_name.finish()), false),
-        ("column_name", Arc::new(column_name.finish()), false),
-        ("storage", Arc::new(storage.finish()), false),
-        ("count", Arc::new(count.finish()), false),
```
Comment (carols10cents) [inline server/src/db/system_tables.rs:371]: ...This chunk columns schema "count" column used to be non-nullable here. Is that right?

File: server/src/db/system_tables.rs [hunk 3]
```diff
@@ -239,22 +301,65 @@ fn from_partition_summaries(partitions: Vec<PartitionSummary>) -> Result<RecordB
...
+}
+
+fn chunk_columns_schema() -> SchemaRef {
+    Arc::new(Schema::new(vec![
+        Field::new("partition_key", DataType::Utf8, false),
+        Field::new("chunk_id", DataType::UInt32, false),
+        Field::new("table_name", DataType::Utf8, false),
+        Field::new("column_name", DataType::Utf8, false),
+        Field::new("storage", DataType::Utf8, false),
+        Field::new("count", DataType::UInt64, true),
```
Comment (carols10cents) [inline server/src/db/system_tables.rs:354]: This chunk columns schema "count" is nullable now, but... see next comment...

**Review by carols10cents — COMMENTED** [review]
I think there's one spot where the nullability of a column changed and I'm not sure whether it was supposed to? Other than that looks great!

File: server/src/db/system_tables.rs (hunk 3 above)
Comment (alamb) [reply server/src/db/system_tables.rs:354]: I double checked and it turns out that `count` can be null. So I guess it was a bug previously. Nice checking.

Specifically, this line: https://github.com/influxdata/influxdb_iox/pull/1452/files#diff-60cd9793ec948d4cac30abf92247edeaa4d413480d7ff30e48591a409af35076L353

```
                count.append_null()?;
```
Actually potentially puts a `NULL` in the `count` column

File: server/src/db/system_tables.rs (hunk 2 above)
Comment (alamb) [reply server/src/db/system_tables.rs:371]: I double checked and I believe this is incorrect

**Review by carols10cents — APPROVED** [review]
I'm good with this now that we're sure `count` nullability is correct!

**Pull request merged**